Training/interface/DataLoader_main.h

```diff
@@ -39,31 +39,37 @@ class CellGrid {
         nCellsEta(_nCellsEta), nCellsPhi(_nCellsPhi), nTotal(nCellsEta * nCellsPhi),
         cellSizeEta(_cellSizeEta), cellSizePhi(_cellSizePhi), cells(nTotal)
     {
-        if(nCellsEta % 2 != 1 || nCellsEta < 1)
+        if(nCellsEta < 1) //(nCellsEta % 2 != 1 || nCellsEta < 1)
             throw std::invalid_argument("Invalid number of eta cells.");
-        if(nCellsPhi % 2 != 1 || nCellsPhi < 1)
+        if(nCellsPhi < 1) //(nCellsPhi % 2 != 1 || nCellsPhi < 1)
             throw std::invalid_argument("Invalid number of phi cells.");
         if(cellSizeEta <= 0 || cellSizePhi <= 0)
             throw std::invalid_argument("Invalid cell size.");
     }
 
-    int MaxEtaIndex() const { return static_cast<int>((nCellsEta - 1) / 2); }
-    int MaxPhiIndex() const { return static_cast<int>((nCellsPhi - 1) / 2); }
-    double MaxDeltaEta() const { return cellSizeEta * (0.5 + MaxEtaIndex()); }
-    double MaxDeltaPhi() const { return cellSizePhi * (0.5 + MaxPhiIndex()); }
+    int MaxEtaIndex() const { return static_cast<int>(nCellsEta / 2); } //{ return static_cast<int>((nCellsEta - 1) / 2); }
+    int MaxPhiIndex() const { return static_cast<int>(nCellsPhi / 2); } //{ return static_cast<int>((nCellsPhi - 1) / 2); }
```

Comment:
This implementation works fine for me. The cell indices are always defined as [-2, -1, 0, 1], and in `FillCellGrid` it skips the cells with the index +2. It didn't look like `MaxDeltaEta/Phi` was used anywhere besides the places where I made modifications. I'll have to look into the `rm_inner_from_outer` option though.
About `FillCellGrid`, I was wondering if there's a reason why the cells are filled in that order, starting from the center bin and going outward? If I change everything so there's only positive integer cell indices like [0, 1, 2, 3, ...], and I start filling cells from one corner to another, will that work just as fine? (Or would I still have to start filling from the center of the grid?)

Reply:
the order in which cells are filled shouldn't matter with the current implementation, because `DataLoader` returns a full matrix. Filling from the centre bin outwards is a legacy of DeepTau v2, where the analogue of the `DataLoader` was returning only non-empty cells and the input grid was built from these cells on the fly during the training. 

ok, if it - and + asymmetry is treated correctly and it is only one place, let's keep it this way to minimise the amount of modification.

Reply:
There are no problems running this with `rm_inner_from_outer`. The bool just filters objects from being filled into the outer grid when they're already in the inner grid, but there's no dependency on the number of cells.

Reply:
yes, for c++ DataLoader it should be fine, but for TF-based data loading it still should be fixed https://github.com/cms-tau-pog/TauMLTools/blob/f0ba2f91a0b0ea5a28838da0c509c9b7786861de/Training/python/2018v1/Training_v0p1.py#L501-L502

Reply:
yes, because in the TF format the grids are already created, and there is no easy way to go back to the particle level. Excluding the inner grid from the outer is mainly intended for debugging/R&D purposes, so this approximation should be fine.